### PR TITLE
Replay protection storage optimization

### PR DESCRIPTION
**Pull request by grarco**
## Describe your changes

Addresses #1566.

Moves the replay protection storage outside of `subspace` to prevent writing the diff keys (still keeps the keys written at the previous height to allow for a rollback). Removes the previous solution to prevent merkelizing the replay protection keys which isn't needed anymore. Finally, implements a small optimization for which only one hash for tx is ever needed in storage (either the wrapper or the inner one, instead of both).

Improves unit tests.

## Indicate on which release or other PRs this topic is based on

v0.23.0

## Checklist before merging to `draft`
- [x] I have added a changelog
- [x] Git history is in acceptable state

### CODE REVIEW

**Review by yito88 — COMMENTED** [review]
Can we call `write_tx_hash` or other functions for replay protection without `wl_storage.write_log`? I mean `wl_storage.write_tx_hash()`.
There are many callers with `wl_storage.write_log` or `wl_storage.storage`. They are a bit confusing because they are not clear when these data are persistent in the storage. (Some bugs happened before by writing data with `wl_storage.storage` directly before the block commit. I think that the direct write to the storage only for testing.

File: apps/src/lib/node/ledger/storage/rocksdb.rs [hunk 1]
```diff
@@ -12,10 +12,12 @@
 //!     epoch can start
 //!   - `next_epoch_min_start_time`: minimum block time from which the next
 //!     epoch can start
+//!   - `replay_protection`: hashes of the processed transactions
 //!   - `pred`: predecessor values of the top-level keys of the same name
 //!     - `tx_queue`
 //!     - `next_epoch_min_start_height`
 //!     - `next_epoch_min_start_time`
+//!     - `replay_protection`
```
[the remark this answers is not in the document]
Comment (grarco) [reply apps/src/lib/node/ledger/storage/rocksdb.rs:20]: Ah no, this is a typo

**Comment by grarco** [comment]
> Can we call `write_tx_hash` or other functions for replay protection without `wl_storage.write_log`? I mean `wl_storage.write_tx_hash()`.
> There are many callers with `wl_storage.write_log` or `wl_storage.storage`. They are a bit confusing because they are not clear when these data are persistent in the storage. (Some bugs happened before by writing data with `wl_storage.storage` directly before the block commit. I think that the direct write to the storage only for testing.

So the keys for replay protection are meant to be written even in case the transaction failed, so the `replay_protection` field of `WriteLog` is treated as the `block_write_log` but I actually never directly write to storage. This is also because I need to write these keys during validation in `process_proposal` and `prepare_proposal` in which the storage is behind an immutable reference. I can reexpose the functions in `WlStorage` if you prefer, to avoid calling them directly from the write log

**Comment by yito88** [comment]
Reexposing the functions in `WlStorage` is better, I think.
In my opinion, WlStorage callers should not choose the write log or the storage (all writes should be stored in the write log) except for the block commit and testing.

**Review by yito88 — APPROVED** [review]

**Pull request merged**